source/memcached-store.ts

```diff
@@ -0,0 +1,171 @@
+	/**
+	 * The number of milliseconds to remember that user's requests.
+	 */
+	expiration!: number
+
+	/**
+	 * The text to prepend to the key.
+	 */
+	prefix!: string
+
+	/**
+	 * The `memcached` client to use.
+	 */
+	client!: MemcachedClient
+
+	/**
+	 * @constructor for `MemcachedStore`.
+	 *
+	 * @param options {Options} - The options used to configure the store's behaviour.
+	 */
+	constructor(options?: Partial<Options>) {
+		this.prefix = options?.prefix ?? 'rl:'
+
+		if (options?.client) {
+			if (
+				typeof options.client.get === 'function' &&
+				typeof options.client.set === 'function' &&
+				typeof options.client.del === 'function' &&
+				typeof options.client.incr === 'function' &&
+				typeof options.client.decr === 'function'
+			)
+				this.client = options.client
+			else throw new Error('An invalid memcached client was passed to store.')
+		} else
+			this.client = new Memcached(
+				options?.locations ?? ['localhost:11211'],
+				options?.config ?? {},
+			)
+	}
+
+	/**
+	 * Method that actually initializes the store.
+	 *
+	 * @param options {RateLimitConfiguration} - The options used to setup the middleware.
+	 *
+	 * @impl
+	 */
+	init(options: RateLimitConfiguration) {
+		this.expiration = options.windowMs / 1000 // [in seconds]
+	}
+
+	/**
+	 * Method to prefix the keys with the given text.
+	 *
+	 * @param key {string} - The key.
+	 *
+	 * @returns {string} - The text + the key.
+	 */
+	prefixKey(key: string): string {
+		return `${this.prefix}${key}`
+	}
+
+	/**
+	 * Method to suffix the keys with `__expiry`, which is the name of the key used
+	 * to store the reset timestamp for that key.
+	 *
+	 * @param key {string} - The key.
+	 *
+	 * @returns {string} - The key + '__expiry'.
+	 */
+	expirySuffix(key: string): string {
+		return `${key}__expiry`
+	}
+
+	/**
+	 * Method to increment a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client.
+	 *
+	 * @returns {IncrementResponse} - The number of hits and reset time for that client.
+	 */
+	async increment(key: string): Promise<IncrementResponse> {
+		const prefixedKey = this.prefixKey(key)
+		const getKey = promisify(this.client.get).bind(this.client)
+		const setKey = promisify(this.client.set).bind(this.client)
+		const incrementKey = promisify(this.client.incr).bind(this.client)
+
+		// Try incrementing the given key. If the key exists, it will increment it
+		// and return the updated hit count.
+		// @ts-expect-error `incrementKey` returns a number or a boolean, not void.
+		// eslint-disable-next-line @typescript-eslint/no-confusing-void-expression
+		let totalHits = (await incrementKey(prefixedKey, 1)) as number | boolean
+		let expiresAt
+
+		if (totalHits === false) {
+			// The increment command failed since the key does not exist. In which case, set the
+			// hit count for that key to 1, and make sure it expires after `window` seconds.
+			await setKey(prefixedKey, 1, this.expiration)
```

Comment:
This technically leaves a window open where, if a bunch of requests came in at the same time, they'd each set the hits to 1, and only the last one would actually be counted. 

To block that, we could call `add` instead of `set` and then call increment if the `add` fails because the key already exists.

Requiring `add` on the `Store` API is a breaking change.

Not sure if it's worth the trouble, but we might just want to document it if we don't make the change.

Reply:
I think this qualifies as a bug, so I went ahead and changed the store to use `add` instead.

> Requiring add on the Store API is a breaking change.

👍🏾